File: pyspark/bigdl/examples/imageframe/inception_validation.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from bigdl.util.common import *

from bigdl.transform.vision.image import *

from bigdl.optim.optimizer import *

from pyspark import SparkContext

from bigdl.nn.layer import *

from optparse import OptionParser

import sys

parser = OptionParser()
parser.add_option("-f", "--folder", type=str, dest="folder", default="",
                  help="url of hdfs folder store the hadoop sequence files")
parser.add_option("--model", type=str, dest="model", default="", help="model path")
parser.add_option("-b", "--batchSize", type=int, dest="batchSize", default=0, help="total batch size")
def get_data(url, sc=None, data_type="val"):
    path = os.path.join(url, data_type)
    return SeqFileFolder.files_to_image_frame(url=path, sc=sc, class_num=1000)

def run(image_path, model_path, batch_size):
    sparkConf = create_spark_conf().setAppName("test_validation")
    sc = get_spark_context(sparkConf)
    init_engine()
    transformer = Pipeline([PixelBytesToMat(), Resize(256, 256), CenterCrop(224, 224),
                            ChannelNormalize(123.0, 117.0, 104.0),
                            MatToTensor(), ImageFrameToSample(input_keys=["imageTensor"],
                                target_keys=["label"])])
    raw_image_frame = get_data(image_path, sc)
    transformed = transformer(raw_image_frame)
    model = Model.loadModel(model_path)
    result = model.evaluate(transformed, int(batch_size), [Top1Accuracy()])
    print "top1 accuray", result[0]

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print "parameters needed : <imagePath> <modelPath> <batchSize>"
    image_path = sys.argv[1]
    model_path = sys.argv[2]
    batch_size = sys.argv[3]
    run(image_path, model_path, batch_size)

File: scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/example/imageclassification/imageFrame/InceptionValidation.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.example.imageclassification.imageFrame

import com.intel.analytics.bigdl.dataset.DataSet.SeqFileFolder
import com.intel.analytics.bigdl.example.loadmodel.ModelValidator.{TestLocalParams, logger, testLocalParser}
import com.intel.analytics.bigdl.nn.Module
import com.intel.analytics.bigdl.optim.{Top1Accuracy, Top5Accuracy}
import com.intel.analytics.bigdl.transform.vision.image.{ImageFrame, ImageFrameToSample, MatToTensor, PixelBytesToMat}
import com.intel.analytics.bigdl.transform.vision.image.augmentation.{CenterCrop, ChannelNormalize, Resize}
import com.intel.analytics.bigdl.utils.Engine
import org.apache.spark.SparkContext
import scopt.OptionParser

object InceptionValidation {
  case class ValidationParam(imageFolder: String = "",
                             modelPath: String = "",
                             batchSize: Int = 32)

  val parser = new OptionParser[ValidationParam]("Inception validation") {
    head("Inception validatio")
    opt[String]('f', "folder")
      .text("where you put the demo image data")
      .action((x, c) => c.copy(imageFolder = x))
      .required()
    opt[String]("modelPath")
      .text("BigDL model")
      .action((x, c) => c.copy(modelPath = x))
      .required()
    opt[Int]('b', "batchSize")
      .text("batch size")
      .action((x, c) => c.copy(batchSize = x))
  }

  def main(args: Array[String]): Unit = {
    parser.parse(args, ValidationParam()).foreach(param => {
      val conf = Engine.createSparkConf()
      conf.setAppName("BigDL ImageFrame API Example")
      val sc = new SparkContext(conf)
      Engine.init
      val imageFrame = SeqFileFolder.filesToImageFrame(param.imageFolder, sc, 1000)
      val model = Module.loadModule[Float](param.modelPath)
      val transformer = PixelBytesToMat() -> Resize(256, 256) ->
        CenterCrop(224, 224) -> ChannelNormalize(123, 117, 104) ->
        MatToTensor[Float]() ->
        ImageFrameToSample[Float](inputKeys = Array("imageTensor"),
          targetKeys = Array("label"))
      imageFrame -> transformer
      val result = model.evaluateImage(
        imageFrame,
        Array(new Top1Accuracy[Float](), new Top5Accuracy[Float]()),
        Some(param.batchSize))
      result.foreach(r => {
        logger.info(s"${ r._2 } is ${ r._1 }")
      })
      sc.stop()
    })
  }
}
